### PR TITLE
[TECH-293] As an MPyL user I want to define my own stage(s)

### PR DESCRIPTION
**Pull request by SamTheisens**
----

### 📕 [GUILD-293](https://vandebron.atlassian.net/browse/GUILD-293) As an MPyL user I want to define my own stage(s) ewaldschrader@vandebron.nl 
At the moment stages are a hardcoded concept in MPyL [steps.py#L60-L85](https://github.com/Vandebron/mpyl/blob/37c7b0c9a7e9eb54dfb05d8822c4e360469855b7/src/mpyl/steps/steps.py#L60-L85) 

While there is potential order dependency between build steps, a fixed order is too limiting. For example, a *security scan_ or _integration test* stage could be run in parallel after the build stage. Defining the execution flow is more at home in the (python) definition of the flow than within the library.

An important aspect to consider is the invalidation logic, which is stage specific (at the moment) [discovery.py#L41-L43](https://github.com/Vandebron/mpyl/blob/e0a8fabcf5c08492f0382bae4df8d5502b45d155/src/mpyl/stages/discovery.py#L41-L43). Perhaps this could be converted to a strategy.

🏗️ Build [9](https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline%20-%20Test/job/PR-333/9/display/redirect) ✅ Successful, started by _Jorg Post_  
🚀 *[cloudfront-service](https://cloudfront-service-333.test.nl/swagger/index.html)*, *example-dagster-user-code*, *job*, *kong-sync*, *[nodeservice](https://nodeservice-333.test.nl/swagger/index.html)*, *ocpp-first*, *ocpp-second*, *[sbtservice](https://sbtservice-333.test.nl/swagger/index.html)*, *sparkJob-first*, *sparkJob-second*  

[GUILD-293]: https://vandebron.atlassian.net/browse/GUILD-293?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/mpyl/stages/discovery.py [hunk 1]
```diff
@@ -5,97 +5,192 @@
...
+        logger, project, stage, changed_file, steps
+    )
+    if invalidated_by_file:
+        output = Output.try_read(project.target_path, stage)
+        return _is_output_invalid(logger, output, change_history, changed_file.revision)
+    return False
 
 
-def output_invalidated(output: Optional[Output], revision_hash: str) -> bool:
+def _is_output_invalid(
```
[the remark this answers is not in the document]
Comment (SamTheisens) [reply src/mpyl/stages/discovery.py:89]: I would rather merge this PR first if you don't mind? It's been open for nearly 3 months.

File: run_properties.yml [hunk 2]
```diff
@@ -1,3 +1,28 @@
+$schema: http://json-schema.org/draft-07/schema#
```
Comment (koeves) [inline run_properties.yml:1]: Why is this schema part of the actual run properties? Shouldn't it be in a run_properties_schema file?

Comment (SamTheisens) [reply run_properties.yml]: Addressed in
[[TECH-293] Move stage definitions to separate schema file](https://github.com/Vandebron/mpyl/pull/333/commits/76022a8c6e672b3e30b552eac917a948195ef79f) and [[TECH-293] Add documentation and health check for stage schema file](https://github.com/Vandebron/mpyl/pull/333/commits/e2ea40720b45db4ec8d39f4f2f9ffd5947affa0c)

**Review by koeves — APPROVED** [review]

File: tests/test_resources/upgrades/run_properties_upgraded.yml [hunk 3]
```diff
@@ -1,4 +1,28 @@
-mpylVersion: 1.3.0
+$schema: http://json-schema.org/draft-07/schema#
```
Comment (koeves) [inline tests/test_resources/upgrades/run_properties_upgraded.yml:1]: This run properties still has the schema in it?

File: mpyl_stages.schema.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+$schema: http://json-schema.org/draft-07/schema#
```
Comment (koeves) [inline mpyl_stages.schema.yml:1]: I would move this file to the schema-s folder in `src/mpyl/schema`

Comment (SamTheisens) [reply mpyl_stages.schema.yml:1]: The schema of the stages needs to be adjustable, so we can't put it in `src/mpyl/schema`

Comment (SamTheisens) [reply mpyl_stages.schema.yml:1]: I fixed the health error here https://github.com/Vandebron/mpyl/pull/333/commits/b73ade4786c2d47df15ac04bd900672337c14a25

**Comment by oschrenk** [comment]
@Jorg88 @koeves For some reason I'm getting notifications from this repository to my personal email address. I'm also able to see this and other repositories. I think it would be best if you remove my permissions.

**Comment by Jorg88** [comment]
> @Jorg88 @koeves For some reason I'm getting notifications from this repository to my personal email address. I'm also able to see this and other repositories. I think it would be best if you remove my permissions.

This is an open source repo, you must be watching it yourself since I don't see you in the access list. Which other repositories can you still see?

**Comment by oschrenk** [comment]
> > @Jorg88 @koeves For some reason I'm getting notifications from this repository to my personal email address. I'm also able to see this and other repositories. I think it would be best if you remove my permissions.
> 
> This is an open source repo, you must be watching it yourself since I don't see you in the access list. Which other repositories can you still see?

Well. That makes sense. I see only public repositories. I guess at some pint I was really interested in mpyl :) 

Sorry for the noise.

**Pull request merged**